docs/frontend-system/building-plugins/06-swappable-components.md

```diff
@@ -0,0 +1,129 @@
+
+```tsx title="in @internal/plugin-example-react"
+import { createSwappableComponent } from '@backstage/frontend-plugin-api';
+
+export const ExampleSwappableComponent = createSwappableComponent({
+  name: 'example',
+
+  // This is a loader for loading the default implementation of the component when there's no overriden
+  // implementation created with `SwappableComponentBlueprint`.
+  // It can be sync like below, but is can also be async like `loader: () => import('./DefaultImplementation').then(m => m.DefaultImplementation)`.
+  loader: () => (props: { name: string }) =>
+    <div>Your name is {props.name}</div>,
+
+  // This is an optional function that can be used to transform the props of the external component.
+  // If it's not provided, the props of the external component will be passed through unchanged.
+  // Typically this wouldn't be used and you don't need to define this initially,
+  // this is a nice API to evolve the Swappable Component internal API without breaking usages of it in existing plugins.
+  transformProps: (props: { lastName: string; firstName: string }) => ({
+    name: `${props.firstName} ${props.lastName}`,
+  }),
+});
+```
+
+## Using a Swappable Component
+
+Using a Swappable Component is just like using any other React Component. The return of `createSwappableComponent` can be rendered directly in your plugin, or any plugins that will be consuming your `-react` package.
+
+```tsx title="in @internal/plugin-example"
+import { ExampleSwappableComponent } from '@internal/plugin-example-react';
+
+<ExampleSwappableComponent firstName="John" lastName="Doe" />;
+```
+
+## Overriding a Swappable Component
+
+In order to override a Swappable Component, you need to create a `SwappableComponentBlueprint` and install it with the `app` plugin.
+There's two different ways to add extensions to the `app` plugin, both are documented below in an example of overriding the `Progress` Swappable Component.
+
+```tsx title="in packages/app/src/App.tsx"
+import {
+  Progress,
+  SwappableComponentBlueprint,
+  createFrontendModule,
+} from '@backstage/frontend-plugin-api';
+import { MyCustomProgress } from './CustomProgress';
+import { createApp } from '@backstage/frontend-defaults';
+import appPlugin from '@backstage/plugin-app';
+
+const app = createApp({
+  features: [
+    // Override an existing extension by ID provided by the app plugin:
+    appPlugin.withOverrides({
+      extensions: [
+        appPlugin
+          .getExtension('component:app/core.components.progress')
+          .override({
+            params: defineParams =>
+              defineParams({
+                component: Progress,
+                loader: () => MyCustomProgress,
+              }),
+          }),
+      ],
+    }),
+    // OR: Add another extension but with the same component ID:
+    appPlugin.withOverrides({
+      extensions: [
+        SwappableComponentBlueprint.make({
+          name: 'core.components.progress',
```

Comment:
Looking at this I realize the `.components.` is a bit redundant. Maybe we could simplify it all the way down to `core-progress`? Both for the name and the component ID. That'd fit alright for plugin libraries too I think, e.g. `scaffolder-template-card`.